tests/std/tests/P2093R14_formatted_output/test.cpp

```diff
@@ -0,0 +1,516 @@
+
+        if constexpr (_Is_ordinary_literal_encoding_utf8()) {
+            // It isn't necessarily well-documented how MultiByteToWideChar() (used internally by
+            // __std_print_to_unicode_console()) replaces invalid code points, except for the fact
+            // that if an invalid code point is encountered, then some amount of characters are
+            // replaced with a single U+FFFD character. So, we instead check that the string written
+            // to the console has at least one U+FFFD character if it is invalid. (The documentation
+            // for the function also implies that if the string provided to MultiByteToWideChar() isn't
+            // empty, then the string created by it also won't be empty.)
+            assert(printed_str.get().empty() || !console_line.empty());
+            assert(contains_replacement_character);
+        }
+
+        // There seems to be some inconsistent behavior with what happens when std::fputs() (used internally
+        // by std::print() et al. when writing to files or when the ordinary literal encoding isn't UTF-8) writes
+        // invalid code sequences to a FILE stream associated with a console. On some systems, it seems that
+        // these characters are dropped; on others, they seem to be replaced with U+FFFD. So, a check like
+        // "assert(!contains_replacement_character)" might pass on some systems and fail on others.
+    };
+
+    // Example UTF-8 Code Sequences from https://www.php.net/manual/en/reference.pcre.pattern.modifiers.php#54805
+
+    // Valid ASCII
+    test_valid_sequence_closure("a");
+
+    // Valid 2 Octet Sequence
+    test_valid_sequence_closure("\xC3\xB1");
+
+    // Invalid 2 Octet Sequence
+    test_invalid_sequence_closure("\xC3\x28");
+
+    // Invalid Sequence Identifier
+    test_invalid_sequence_closure("\xA0\xA1");
+
+    // Valid 3 Octet Sequence
+    test_valid_sequence_closure("\xE2\x82\xA1");
+
+    // Invalid 3 Octet Sequence (in 2nd Octet)
+    test_invalid_sequence_closure("\xE2\x28\xA1");
+
+    // Invalid 3 Octet Sequence (in 3rd Octet)
+    test_invalid_sequence_closure("\xE2\x82\x28");
+
+    // Skipped: 0xF0 0x90 0x8C 0xBC (MultiByteToWideChar() correctly finds that this is a valid
+    // UTF-8 code point, but the call to WriteConsoleW() seems to erroneously replace it with U+FFFD.)
+
+    // Invalid 4 Octet Sequence (in 2nd Octet)
+    test_invalid_sequence_closure("\xF0\x28\x8C\xBC");
+
+    // Invalid 4 Octet Sequence (in 3rd Octet)
+    test_invalid_sequence_closure("\xF0\x90\x28\xBC");
+
+    // Invalid 4 Octet Sequence (in 4th Octet)
+    test_invalid_sequence_closure("\xF0\x28\x8C\x25");
+}
+
+void test_invalid_code_points_file() {
+    // Unlike for the console API when the ordinary literal encoding is UTF-8, invalid code points shouldn't
+    // be replaced when writing to a file.
+    FILE* temp_file_stream;
+
+    {
+        const errno_t tmpfile_result = tmpfile_s(&temp_file_stream);

```

Comment:
We've had trouble with functions that generate temporary file names in the user-global temp directory, due to collisions during heavily concurrent test runs - see #2210. We should use the same technique as `tests/tr1/include/temp_file_name.h` here, I think.